### PR TITLE
chore: OP modules improvements

### PR DESCRIPTION
**Pull request by varasev**
Fixes https://github.com/blockscout/blockscout/issues/11086, fixes https://github.com/blockscout/blockscout/issues/11101, closes https://github.com/blockscout/blockscout/issues/11077.

## Motivation

This PR does the following things:
- adds a new socket channel with `optimism:new_batch` topic and `new_optimism_batch` event for the frontend (see details below)
- replaces `optimism_deposits:new_deposits` socket topic (and `deposits` event) with `optimism:new_deposits` topic (and `new_optimism_deposits` event) for consistency. The payload remains the same
- adds fallback envs to use them when the `SystemConfig` contract is obsolete: `INDEXER_OPTIMISM_L1_PORTAL_CONTRACT`, `INDEXER_OPTIMISM_L1_START_BLOCK`, `INDEXER_OPTIMISM_L1_BATCH_INBOX`, `INDEXER_OPTIMISM_L1_BATCH_SUBMITTER`
- adds `INDEXER_OPTIMISM_L1_ETH_GET_LOGS_RANGE_SIZE` and `INDEXER_OPTIMISM_L2_ETH_GET_LOGS_RANGE_SIZE` env variables to be able to define request limits for `eth_getLogs` requests (instead of the hardcoded values)
- removes `INDEXER_OPTIMISM_L1_DEPOSITS_BATCH_SIZE` env variable in favor of `INDEXER_OPTIMISM_L1_ETH_GET_LOGS_RANGE_SIZE`
- adds `INDEXER_OPTIMISM_BLOCK_DURATION` env variable to be able to define block time (which is 2 seconds by default for OP chains)
- fixes the `Deposits` module (now it doesn't use `eth_newFilter` and `eth_getFilterChanges`)
- adds `timeout: :infinity` for some db select requests
- adds support of transactions without `to` field in `eth_getTransactionByHash` RPC response
- adds logic to find the last known block after reorg and restart of the instance (only for OP fetcher modules)
- hardcodes `INDEXER_BEACON_BLOB_FETCHER_*` env variables for `Indexer.Fetcher.Optimism.TransactionBatch` module (see https://github.com/blockscout/blockscout/issues/11077)

## New/updated sockets

https://github.com/blockscout/frontend needs to be updated to read the new socket channel (`optimism:new_batch` topic and `new_optimism_batch` event).

The payload for that looks as follows:

```
"batch": {
  "internal_id": 123,
  "l1_timestamp": "2024-01-29T21:31:35.000000Z",
  "transaction_count": 237
}
```

Also, `optimism_deposits:new_deposits` topic was replaced with `optimism:new_deposits` (and the `deposits` event with `new_optimism_deposits` correspondingly). The payload remains the same as before.

## New env variables

- `INDEXER_OPTIMISM_L1_PORTAL_CONTRACT` (as fallback when `SystemConfig` is old)
- `INDEXER_OPTIMISM_L1_START_BLOCK` (as fallback when `SystemConfig` is old)
- `INDEXER_OPTIMISM_L1_BATCH_INBOX` (as fallback when `SystemConfig` is old)
- `INDEXER_OPTIMISM_L1_BATCH_SUBMITTER` (as fallback when `SystemConfig` is old)
- `INDEXER_OPTIMISM_L1_ETH_GET_LOGS_RANGE_SIZE` (optional)
- `INDEXER_OPTIMISM_L2_ETH_GET_LOGS_RANGE_SIZE` (optional)
- `INDEXER_OPTIMISM_BLOCK_DURATION` (optional)

See their descriptions in https://github.com/blockscout/docs/pull/337

## Fix for `Indexer.Fetcher.Optimism.Deposit`

Now the module doesn't use `eth_newFilter` and `eth_getFilterChanges` JSON RPC requests as they work incorrectly on Nethermind (see https://github.com/NethermindEth/nethermind/issues/7751).

## Support transactions without `to` field

Contract-creating transactions don't have the `to` field in `eth_getTransactionByHash` response sent by `reth` RPC node. This fix eliminates the error described in https://github.com/blockscout/blockscout/issues/11086 and https://github.com/blockscout/blockscout/issues/11101.

## Finding correct block number to start from after reorg

OP fetchers were improved in the part of reorgs handling: now OP fetcher will try to find nearest block from which it can continue to work after reorg and restarting the instance. Previous implementation just threw an error in logs and stopped the fetcher.

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs). https://github.com/blockscout/docs/pull/337
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables).
- [ ] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [ ] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

**Review by vbaranov — COMMENTED** [review]
1. 
> replaces optimism_deposits:new_deposits socket topic (and deposits event) with optimism:new_deposits topic (and new_optimism_deposits event) for consistency. The payload remains the same

This breaking change requires to synchronize releases between backend and frontend which is complicated thing. Can we temporarily keep compatibility with the current version of the frontend? I'd suggest implementing new events and keep old events until frontend will not bind to new ones. Such a multi-step process gives us more flexibility in backend release publishing and more time to find possibly broken functionality.

2. Seems, web application tests on Filecoin chain type are failed.

**Comment by varasev** [comment]
> This breaking change requires to synchronize releases between backend and frontend which is complicated thing. Can we temporarily keep compatibility with the current version of the frontend? I'd suggest implementing new events and keep old events until frontend will not bind to new ones. Such a multi-step process gives us more flexibility in backend release publishing and more time to find possibly broken functionality.

Yes, I returned those socket topic/event in https://github.com/blockscout/blockscout/pull/11073/commits/c2c52dde935f4d52da5d6fa13f06903a96e89e33

File: apps/indexer/lib/indexer/helper.ex [hunk 1]
```diff
@@ -178,6 +178,10 @@ defmodule Indexer.Helper do
     repeated_call(&json_rpc/2, [req, json_rpc_named_arguments], error_message, retries)
   end
 
+  def finite_retries_number do
```
[the remark this answers is not in the document]
Comment (varasev) [reply apps/indexer/lib/indexer/helper.ex:186]: Done in https://github.com/blockscout/blockscout/pull/11073/commits/c2c52dde935f4d52da5d6fa13f06903a96e89e33

File: apps/indexer/lib/indexer/fetcher/optimism.ex [hunk 2]
```diff
@@ -380,4 +384,66 @@ defmodule Indexer.Fetcher.Optimism do
...
+
+    ## Returns
+    - A tuple `{last_block_number, last_transaction_hash, last_transaction}` where
+      `last_block_number` is the last block number found in the corresponding table (0 if not found),
+      `last_transaction_hash` is the last transaction hash found in the corresponding table (nil if not found),
+      `last_transaction` is the transaction info got from the RPC (nil if not found or not needed).
+    - A tuple `{:error, message}` in case the `eth_getTransactionByHash` RPC request failed.
+  """
+  @spec get_last_item(:L1 | :L2, function(), function(), EthereumJSONRPC.json_rpc_named_arguments() | nil) ::
+          {non_neg_integer(), binary() | nil, map() | nil} | {:error, any()}
```
[the remark this answers is not in the document]
Comment (varasev) [reply apps/indexer/lib/indexer/fetcher/optimism.ex:413]: Yes, we return the tuple here. I renamed `error_message` to just `error` in https://github.com/blockscout/blockscout/pull/11073/commits/c2c52dde935f4d52da5d6fa13f06903a96e89e33

**Comment by varasev** [comment]
> 2\. Seems, web application tests on Filecoin chain type are failed.

I think, this is not related to the current changes, but I'll try to rebase onto master and will see

**Comment by vbaranov** [comment]
> > This breaking change requires to synchronize releases between backend and frontend which is complicated thing. Can we temporarily keep compatibility with the current version of the frontend? I'd suggest implementing new events and keep old events until frontend will not bind to new ones. Such a multi-step process gives us more flexibility in backend release publishing and more time to find possibly broken functionality.
> 
> Yes, I restored those socket topic/event in [c2c52dd](https://github.com/blockscout/blockscout/commit/c2c52dde935f4d52da5d6fa13f06903a96e89e33)

@varasev shouldn't you keep corresponding channel apps/block_scout_web/lib/block_scout_web/channels/optimism_deposit_channel.ex for compatibility as well?

**Comment by varasev** [comment]
> @varasev shouldn't you keep corresponding channel apps/block_scout_web/lib/block_scout_web/channels/optimism_deposit_channel.ex for compatibility as well?

You mean the `deposits` handler from here? https://github.com/blockscout/blockscout/blob/50da37f33601b1b8192570497a594f6d0bebb17f/apps/block_scout_web/lib/block_scout_web/channels/optimism_deposit_channel.ex

It's here now: https://github.com/blockscout/blockscout/blob/c2c52dde935f4d52da5d6fa13f06903a96e89e33/apps/block_scout_web/lib/block_scout_web/notifiers/optimism.ex#L27-L29

**Comment by vbaranov** [comment]
> Endpoint.broadcast("optimism_deposits:new_deposits", "deposits", %{

@varasev let's mark with `todo` comments parts of the code which are temporarily added for backward compatibility and conditions when those parts should be removed.

**Comment by varasev** [comment]
> @varasev let's mark with `todo` comments parts of the code which are temporarily added for backward compatibility and conditions when those parts should be removed.

Done in https://github.com/blockscout/blockscout/pull/11073/commits/f5b14a40427c80bc6d3c92f2694f2b37fbafaa27

File: apps/indexer/lib/indexer/fetcher/optimism/deposit.ex [hunk 3]
```diff
@@ -63,266 +49,100 @@ defmodule Indexer.Fetcher.Optimism.Deposit do
...
+
+          deposit_events = prepare_events(result, transaction_type, json_rpc_named_arguments)
+
+          {:ok, _} =
+            Chain.import(%{
+              optimism_deposits: %{params: deposit_events},
+              timeout: :infinity
+            })
+
+          Publisher.broadcast(%{new_optimism_deposits: deposit_events}, :realtime)
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/fetcher/optimism/deposit.ex:105]: Seems, there is no backward compatibility with the current frontend here. We should also broadcast the old event as well.

**Comment by varasev** [comment]
> > You mean the `deposits` handler from here? https://github.com/blockscout/blockscout/blob/50da37f33601b1b8192570497a594f6d0bebb17f/apps/block_scout_web/lib/block_scout_web/channels/optimism_deposit_channel.ex
> > It's here now:
> > https://github.com/blockscout/blockscout/blob/c2c52dde935f4d52da5d6fa13f06903a96e89e33/apps/block_scout_web/lib/block_scout_web/notifiers/optimism.ex#L27-L29
> 
> Yeah, I meant that. But how about handling of the old event in `notifier.ex`? I don't see handler for the old event. Also, I don't see the definition of the old channel in `user_socket.ex` and `user_socket_v2.ex`.

Ah, forgot to remove `:` in the `UserSocketV2`. Fixed in https://github.com/blockscout/blockscout/pull/11073/commits/de147bc31acb56ba682449d823a46168b527913e. The event is handled here: https://github.com/blockscout/blockscout/blob/de147bc31acb56ba682449d823a46168b527913e/apps/block_scout_web/lib/block_scout_web/notifiers/optimism.ex#L30-L32

Regarding the `user_socket.ex` - do we really need that for Optimism? E.g., there is no Arbitrum handler there.

Don't worry, I locally tested the old and new socket topics with `PieSocket Websocket Tester` - everything works fine

File: apps/indexer/lib/indexer/fetcher/optimism/deposit.ex (hunk 3 above)
Comment (varasev) [reply apps/indexer/lib/indexer/fetcher/optimism/deposit.ex:105]: This line is responsible for the internal broadcasting of the `new_optimism_deposits ` notifier event. It's caught by the handler in `BlockScoutWeb.Notifiers.Optimism` and then `optimism:new_deposits` and `optimism_deposits:new_deposits` topics are broadcasted to the socket

File: apps/block_scout_web/lib/block_scout_web/channels/user_socket_v2.ex [hunk 4]
```diff
@@ -16,6 +15,7 @@ defmodule BlockScoutWeb.UserSocketV2 do
 
   case Application.compile_env(:explorer, :chain_type) do
     :arbitrum -> channel("arbitrum:*", BlockScoutWeb.ArbitrumChannel)
+    :optimism -> channel("optimism*", BlockScoutWeb.OptimismChannel)
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/channels/user_socket_v2.ex:18]: I believe we should add a to-do comment as well here in order to return `:` in the future once we eliminate the old event. wdyt?

Comment (varasev) [reply apps/block_scout_web/lib/block_scout_web/channels/user_socket_v2.ex:19]: Yes, done in https://github.com/blockscout/blockscout/commit/d89f2c93111fa401ea3d5b519f09546ee9084331

**Review by vbaranov — APPROVED** [review]

**Pull request merged**